diff --git a/include/configs/rpi.h b/include/configs/rpi.h
index c499b45..8967500 100644
--- a/include/configs/rpi.h
+++ b/include/configs/rpi.h
@@ -90,6 +90,8 @@
 #define CONFIG_ENV_VARS_UBOOT_CONFIG
 #define CONFIG_SYS_LOAD_ADDR		0x1000000
 #define CONFIG_PREBOOT			"usb start"
+#define CONFIG_BOOTCOUNT_LIMIT
+#define CONFIG_BOOTCOUNT_ENV

 /* Shell */
 #define CONFIG_CMDLINE_EDITING
@@ -156,6 +158,15 @@

 #define CONFIG_EXTRA_ENV_SETTINGS \
 	"dhcpuboot=usb start; dhcp u-boot.uimg; bootm\0" \
+   "bootlimit=2\0" \
+	"bootcount=0\0" \
+	"upgrade_available=0\0" \
+	"kuksa_part_A=2\0"\
+	"kuksa_part_B=3\0"\
+	"kuksa_start=setenv expand_bootargs \"setenv bootargs \\\\\"${bootargs}\\\\\"\"; run expand_bootargs; setenv expand_bootargs \0" \
+	"kuksa_init=if printenv kuksa_root ; then echo configuration is set up ; else setenv kuksa_root ${kuksa_part_A}; saveenv; fi\0" \
+	"altbootcmd=if test ${kuksa_root} = ${kuksa_part_A}; then setenv kuksa_root ${kuksa_part_B}; else setenv kuksa_root ${kuksa_part_A};fi;setenv upgrade_available 0; setenv bootcount 0; saveenv; run bootcmd\0" \
+	"kuksa_recover= if test ${upgrade_available} = 1; then reset; fi\0" \
 	ENV_DEVICE_SETTINGS \
 	ENV_MEM_LAYOUT_SETTINGS \
 	BOOTENV
